### DISCUSSION TITLE
Is it possible to create API Routes based on discriminator values?

### DISCUSSION THREAD

**Discussion by egorksv** [fastapi/fastapi#11196] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@api.post("/")
def root(
        webhook: Union
        [
            WebhookUserRequest,
            WebhookEnrollmentRequest
        ] = Body(..., discriminator="resource"),
):
    if webhook.resource == "user":
        thinkific_user_sync.sync_user(webhook.payload, thinkific_site)
        return {"message": "User Synced"}
    elif webhook.resource == "enrollment":
        thinkific_enrollment_sync.sync_enrollment(webhook.payload, thinkific_site)
        return {"message": "Enrollment Synced"}



class WebhookRequest(BaseModel):
    id: str
    action: str
    tenant_id: str
    tenant_global_id: str = None
    created_at: datetime = None
    timestamp: int = None


class WebhookUserRequest(WebhookRequest):
    resource: Literal["user"] = "user"
    payload: ThinkificUser


class WebhookEnrollmentRequest(WebhookRequest):
    resource: Literal["enrollment"] = "enrollment"
    payload: ThinkificEnrollment
```

### Description

I have external system (thinkific) that sends webhooks as POST requests to the same URL, passing "resource" and "action" as part of the json body. I have no problems mapping that as a discriminated union, however, this essentially creates a "second-level hardcoded router" which creates a lot of maintainability issues.

Is there a way (maybe using customised APIRouter) to map these onto different functions?
Something akin to this:

```
@api.post("/")
def root(
        webhook: WebhookUserRequest
):
  pass

@api.post("/")
def root(
        webhook: WebhookEnrollmentRequest
):
  pass

```

### Operating System

Linux

### FastAPI Version

"0.109.0"

### Pydantic Version

"2.6.0"

### Python Version

3.11.7

**Comment by chrisK824** [comment, marked as answer]
I don't think that's possible. By the time the route is defined in your FastAPI app, the request payload is not known, cannot be known, so you simply cannot define a route based on incoming requests. It's like the request should determine itself where to go based on what data the caller put inside it. And you can't have two handlers for one route anyway. What stops you from having that simple match case there and keep the rest of the code in separate standalone functions that behave like endpoints ?

**Comment by YuriiMotov** [comment]
I agree with @chrisK824.

If you have wide range of such cases and all of them follow the same naming pattern, you can use this approach:
```
    module = globals()[f"thinkific_{webhook.resource}_sync"]
    func = getattr(module, f"sync_{webhook.resource}")

    func(webhook.payload, thinkific_site)
    return {"message": f"{webhook.resource.title()} Synced"}
```

**Reply by egorksv** [reply]
Yeah there are many solutions that fall under "just write your own router" category.
I was thinking more in terms of "we already have perfectly awesome router in fastapi, any way to delegate to it?"
